pkg/kubelet/prober/manager.go

```diff
@@ -0,0 +1,141 @@
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+	"k8s.io/kubernetes/pkg/kubelet/status"
+	"k8s.io/kubernetes/pkg/types"
+)
+
+// Manager manages pod probing. It creates a probe "worker" for every container that specifies a
+// probe (AddPod). The worker periodically probes its assigned container and caches the results. The
+// manager usse the cached probe results to set the appropriate Ready state in the PodStatus when
+// requested (UpdatePodStatus).
+// TODO: Move liveness probing out of the runtime, to here.
+type Manager interface {
+	// AddPod creates new probe workers for every container probe. This should be called for every
+	// pod created.
+	AddPod(pod *api.Pod)
+
+	// RemovePod handles cleaning up the removed pod state, including terminating probe workers and
+	// deleting cached results.
+	RemovePod(pod *api.Pod)
+
+	// UpdatePodStatus modifies the given PodStatus with the appropriate Ready state for each
+	// container based on container running status, cached probe results and worker states.
+	UpdatePodStatus(types.UID, *api.PodStatus)
+}
+
+type manager struct {
+	// Caches the results of readiness probes.
+	readinessCache *readinessManager
+
+	// Map of active workers for readiness
+	readinessProbes map[containerPath]*worker
+	// Lock for accessing & mutating readinessProbes
+	workerLock sync.RWMutex
+
+	// The statusManager cache provides pod IP and container IDs for probing.
+	statusManager status.Manager
+
+	// prober executes the probe actions.
+	prober Prober
+
+	// Default period for workers to execute a probe.
+	defaultProbePeriod time.Duration
+}
+
+func NewManager(
+	defaultProbePeriod time.Duration,
+	statusManager status.Manager,
+	prober Prober) Manager {
+	return &manager{
+		defaultProbePeriod: defaultProbePeriod,
+		statusManager:      statusManager,
+		prober:             prober,
+		readinessCache:     newReadinessManager(),
+		readinessProbes:    make(map[containerPath]*worker),
+	}
+}
+
+// Key uniquely identifying containers
+type containerPath struct {
+	podUID        types.UID
+	containerName string
+}
+
+func (m *manager) AddPod(pod *api.Pod) {
+	m.workerLock.Lock()
+	defer m.workerLock.Unlock()
+
+	key := containerPath{podUID: pod.UID}
+	for _, c := range pod.Spec.Containers {
+		key.containerName = c.Name
+		if _, ok := m.readinessProbes[key]; ok {
+			glog.Errorf("Readiness probe already exists! %v - %v",
+				kubecontainer.GetPodFullName(pod), c.Name)
+			return
+		}
+		if c.ReadinessProbe != nil {
+			m.readinessProbes[key] = m.newWorker(pod, c)
+		}
+	}
+}
+
+func (m *manager) RemovePod(pod *api.Pod) {
+	m.workerLock.RLock()
+	defer m.workerLock.RUnlock()
+
+	key := containerPath{podUID: pod.UID}
+	for _, c := range pod.Spec.Containers {
+		key.containerName = c.Name
+		if worker, ok := m.readinessProbes[key]; ok {
+			close(worker.stop)
+		}
+	}
+}
+
+func (m *manager) UpdatePodStatus(podUID types.UID, podStatus *api.PodStatus) {
```

Comment:
Yes, in the interest of keeping this PR size down, I opted to address this in a follow up change. The two options for how to address this are to update the status manager directly (by adding a SetReadiness method on the status manager), or to provide a "statusSync" channel which triggers a status update on readiness changes. The former might be prefered for readiness, while the latter might be better for liveness.